Decrease indentation when line begins with a right paren

d-torrance opened the pull request:
As suggested in #77, we decrease the indentation when indenting a line that opens with a `)`, `]`, or `}`.  We also revive the unused `M2-electric-right-brace` function so that we re-indent automatically when typing one of these characters.

![Peek 2024-12-19 15-05](https://github.com/user-attachments/assets/f664fcdb-487e-4584-817d-4c60199f939b)

d-torrance committed:
Adjust indentation when lines begin with right parens

If the first character in the current line is a right paren, then we
decrease the indentation.

Also delete the now-unused function M2-next-line-indent-amount.

d-torrance committed:
Revive M2-electric-right-brace

Its keybinding to "}" has been commented out for ages, but it will be
very useful for re-indenting lines that begin with a right paren.

In addition to "}", we also bind it to ")" and "]".

mahrud commented:
Amazing! In lieu of a review I will test drive this for a while and report back!

mahrud commented:
So far happy with this. As long as nobody else is opposed to this change, I'm happy to merge this.

mahrud commented:
Things are working okay, but I'm getting pretty annoyed when editing older code.

For instance say I start with an old code
```m2
f = x -> (
    2 * x
    )
<-- pointer here
```
Then I press enter to make a new line and add a new function, auto indent gives me:
```m2
f = x -> (
    2 * x
    )

    g = x -> (
    )
```
On the other hand, say I start fresh with:
```m2
f = x -> (
    x = 2 * x;) <-- pointer here before the parenthesis
```
Then I press enter to make a new line and add a new statement, this happens:
```m2
f = x -> (
    x = 2 * x;
return x) <-- pointer end up here, before the parenthesis
```
Now if I press tab I get:
```m2
f = x -> (
    x = 2 * x;
return x    ) <-- pointer end up here, before the parenthesis
```
Which isn't what I wanted ... I need to jump to beginning of the line then tab to indent it correctly.

Not totally sure if these are fixable, but you might know of another option.

d-torrance commented:
I finally played around with this some more.  I think I've addressed both issues!  In particular:

* The indentation is now calculated based on the paren depth rather than just the paren change in the previous line and the current indentation.  So it shouldn't get messed up even if nearby code isn't indented properly.
* <kbd>Tab</kbd>'ing when to the right of non-whitespace text now re-indents the line.

mahrud commented:
Great, I'll give it a try!

mahrud commented:
By the way, it would be good to bring this up in Tulane.

d-torrance commented:
It would be. Will you be there? I won't

mikestillman commented:
I'll be there.  What all should I say about it?  When will it take effect for people's code?  Probably for emacs, and when they install the next version?

mikestillman commented:
I need to try this, but it sounds good!

d-torrance commented:
@mikestillman - If this is merged, then it would appear with the next release in May.

I put together a short video demonstrating the proposed changes: https://youtu.be/o28KrBMWa2w

d-torrance commented:
One big disadvantage of the latest change (determining indentation based on the global paren depth rather than just locally looking at the previous line's indentation and paren change) is that it messes SimpleDoc docstrings up.  Since they're not surrounded by any parentheses, it thinks that all the lines should start at the very beginning, with no indentation.

An easy workaround is to just use <kbd>C</kbd>-<kbd>j</kbd> instead of <kbd>RET</kbd>, but this may be a deal breaker.

d-torrance commented:
Actually, the SimpleDoc issue might have been a problem even in the first iteration of this PR...

I'm thinking of adding an additional "are we inside a ///-delimited string?" function (maybe using font-lock), and if so, adjust the indentation rules accordingly.  Probably nothing too fancy -- maybe <kbd>Tab</kbd> gives us 2 spaces and we don't re-indent automatically.

mahrud commented:
Do you happen to have a fix for the SimpleDoc issue? No worries if you haven't had a chance to look at it yet.

d-torrance commented:
I've played with it a bit, but nothing working yet.

mikestillman commented:
I am generally a person who is used to the older way of doing this (ie. indent to below last lines, not one level out).  I need to play with it.  In some sense this is ok, and seems more bullet-proof.  But I am a skeptic who needs a bit of convincing perhaps...

Here is something I don't know how easily to do with the new indentation rules.  If I type n//2, then go to the next line, it reindents to under the if.  Not sure what I think about that.  I guess if I want to always use parens for if-then-else, that might be ok.  But otherwise, I have to move off the line first, and make sure I don't re-indent the region.  What do you think?

```m2
f = (n) -> (
    if even n then
        n // 2
```

mikestillman commented:
I found your video useful too...!

d-torrance commented:
Yeah, I see what you're saying.  I usually use parentheses every time I want the indentation to increase, but I see the appeal of omitting them.

mikestillman commented:
this is a problem also in the current M2 file, in that it won't indent it.  However, it doesn't move it on you either.  I'm not sure what people will think about that.  It is nice to do it automatically, but I'm worried about what users will think.  They might be fine, they might not.  By the way, at Tulane, I would say most people appeared to be using vscode.  So I'm not sure if they would be good ones to ask about an emacs thing.  We should have some other people who use it alot try it.  Maybe Greg, Anton, David.  I'll ask Dan too.

mahrud commented:
> By the way, at Tulane, I would say most people appeared to be using vscode. So I'm not sure if they would be good ones to ask about an emacs thing. We should have some other people who use it alot try it. Maybe Greg, Anton, David. I'll ask Dan too.

The point of this PR is to bring the indentation behavior in Emacs in-line with what vscode already does, and avoid having two editors that indent code differently. The alternative would be to figure out how to change indentation rules in the vscode extension.

mahrud commented:
@d-torrance FYI I was too frustrated with the current version of this PR unindenting my tests left and right, so I switched back to the older version of this PR, which in comparison was much less frustrating. If possible, I think the following is a decent compromise, if it's doable:
- on a given line, if the previous line is empty, use global indentation.
- otherwise inherit indentation baseline from the previous line.

d-torrance commented:
I'll play around with this some more soon.

FYI -- by default, VS Code automatically closes parens, e.g., if you type "(", it automatically inserts a closing ")".  You can replicate this in Emacs using `M-x electric-pair-mode`.  Maybe this is worth putting in doc somewhere.

mikestillman commented:
> @d-torrance FYI I was too frustrated with the current version of this PR unindenting my tests left and right, so I switched back to the older version of this PR, which in comparison was much less frustrating. If possible, I think the following is a decent compromise, if it's doable:
> 
> * on a given line, if the previous line is empty, use global indentation.
> * otherwise inherit indentation baseline from the previous line.

I also found the changing indentation in doc nodes to be hard to work with, so I had to revert back to the old behavior too.

I personally don't like filling in closing brace too when e.g. `(` is pressed, but for those who are used to it, it might be nice to have that as an option.  For me, I would not want that to be the default.

d-torrance committed:
Adjust TAB behavior when not in the front of the current line.

Previously, we added additional tabs/spaces at the point, but now we
re-indent the line, which is the usual behavior in many other major
modes.

We add a new helper function, M2-indent-line, which we can also use to
simplify M2-electric-right-brace.
